src/server/string_family.cc

```diff
@@ -1420,39 +1420,48 @@ void MGetGeneric(CmdArgList args, CommandContext* cmnd_cntx,
   // wait for all tiered reads to finish and check for errors
   tiering_bc->Wait();
   if (auto err = std::move(tiering_err).Destroy(); err)
-    return builder->SendError(err.message());
-
-  // reorder shard results back according to argument order
-  absl::FixedArray<optional<GetResp>, 8> res(args.size());
-  ReorderShardResults(absl::MakeSpan(mget_resp), cmnd_cntx->tx, is_memcache, absl::MakeSpan(res));
-
-  SinkReplyBuilder::ReplyScope scope(builder);
-  if (is_memcache) {
-    // TODO: implement deferred support for MGET, which prevents us from running SET,GET in async
-    // pipeline.
-    DCHECK(!cmnd_cntx->IsDeferredReply());
-    auto* rb = static_cast<MCReplyBuilder*>(builder);
-    DCHECK(dynamic_cast<CapturingReplyBuilder*>(builder) == nullptr);  // memcache is never squashed
-    for (const auto& entry : res) {
-      if (entry) {
-        // TODO: we will need to pass the whole response to ParsedCommand to support
-        // capturing MC replies.
-        rb->SendValue(cmnd_cntx->mc_command()->cmd_flags, entry->key, entry->value, 0,
-                      entry->mc_flag, cmnd_cntx->mc_command()->replies_cas_token());
-      } else {
-        cmnd_cntx->SendMiss();
+    return cmnd_cntx->SendError(err.message());
+
+  size_t arg_len = args.size();
+
+  unique_ptr<optional<GetResp>[]> mget_results(new optional<GetResp>[arg_len]);
+  ReorderShardResults(absl::MakeSpan(mget_resp.get(), shard_set->size()), cmnd_cntx->tx,
+                      absl::MakeSpan(mget_results.get(), arg_len));
+
+  if (cmnd_cntx->mc_command()) {
+    // We can not access `args` inside the lambda below if it runs asynchronously.
+    // Access to arguments must be done via cmnd_cntx.
```

Comment:
why can't you just capture it by value? if there's two ways to do a thing it should be simpler

Reply:
because currently  `args` are not necessary span around `cmnd_cntx`.  That's something that we need to fix in the long run but for now args can reference other argument storages. We have many places where we invoke commands, so args that we pass there wrap different backed argument than those stored inside `cmnd_cntx`